### ISSUE TITLE
Inconsistent input processing, error reporting

### ISSUE THREAD

**Issue by flexoron** [infradig/trealla#174]
Input is 1/2/3/

?- X is 1/2/3/.
uncaught exception: error(existence_error(procedure,'/.'),'/.'/0)
false.

?- X is 1/2/3/ .
Error: syntax error

?- assertz(g(1/2/3/)).
true.
?- g(M).
M = 1/2/3. 
the last slash was eaten? why?

**Comment by flexoron** [comment]
and
?- assertz(g(/1/2/3/)).
true.
?- g(M).
uncaught exception: error(existence_error(procedure,g(_2)),g/1)
false.

**Comment by flexoron** [comment]
ah wrong version error report ok
Trealla Prolog (c) Infradig 2020, v1.7.19
?- X is 1/2/3/.
Error: syntax error
?- X is 1/2/3/ .
Error: syntax error

Are these errors?
?- assertz(g(1/2/3/)).
true.
?- g(M).
M = 1/2/3.
?- halt.
You've to halt to get this below

Trealla Prolog (c) Infradig 2020, v1.7.19
?- assertz(g(/1/2/3/)).
true.
?- g(M).
uncaught exception: error(existence_error(procedure,g(_2)),g/1)
false.

**Comment by flexoron** [comment]
why are some quoted?
Trealla Prolog (c) Infradig 2020, v1.7.19
?- assertz(g(/)).
true.
?- assertz(g(//)).
true.
?- assertz(g(///)).
true.
?- g(M).
M = / 
; M = // 
; M = '///'.
?- assertz(g(////)).
true.
?- g(M).
M = / 
; M = // 
; M = '///' 
; M = '////'.
?-

**Comment by pmoura** [comment]
> Hey @pmoura , the parser improvements have picked up a syntax error in logtalk3/adapters/trealla.pl
> 
> ```
> '$lgt_prolog_meta_predicate'(task(_, _, _, _, _, _, _), task(6, *, *, *, *, *. *), predicate).
> ```
> 
> note the dot instead of a comma!

Thanks for reporting. Fixed in the Logtalk git version.

P.S. Eventually these clauses will not be required as the necessary data will be available by simply calling the `predicate_property/2` predicate (see e.g. the adapter file for GNU Prolog).

**Comment by infradig** [comment]
I'll have a look at doing meta_predicate & upgrading predicate_propery.

**Comment by infradig** [comment]
Well / & // are operators so don't get quoted. So /// etc are getting
treated as atoms and hence quoted. Will look into it.

**Comment by infradig** [comment]
What are the definitions for task/1-7 about?

**Comment by pmoura** [comment]
> What are the definitions for task/1-7 about?

The Logtalk compiler requires knowing when a call to a backend built-in predicate inside an object (or a category) is actually a call to a meta-predicate so that the call is compiled correctly. Ideally, the compiler could simply call `predicate_property/2` to find out if a predicate is (1) a built-in predicate and (2) a meta-predicate (accessing in that case its meta-predicate template). The Logtalk internal `'$lgt_prolog_meta_predicate'/3` predicate is a fallback, which is currently used in the Trealla Prolog adapter file as we get e.g. for `task/3`:

```text
?- predicate_property(task(_, _, _), built_in).
true.
?- predicate_property(task(_, _, _), meta_predicate(M)).
uncaught exception: error(type_error(atom,meta_predicate(_5)),predicate_property/2)
false.
```

Btw, `predicate_property/2` should be able to enumerate by backtracking all properties of a predicate when called with an unbound second argument (the first argument is required to be bound).

**Comment by infradig** [comment]
> Btw, predicate_property/2 should be able to enumerate by backtracking all
properties of a predicate when called with an unbound second argument (the
first argument is required to be bound).

That will have to wait a while.

**Comment by infradig** [comment]
BTW, you can currently get the meta_predicate/1 info

$ cat x.pl
:- meta_predicate(meta(0, +)).

$ tpl x.pl
Trealla Prolog (c) Infradig 2020, v1.7.20-2-g8d9c0
?- listing.
:- meta_predicate(meta(0,+)).
true.
?- (:- meta_predicate(X)).
X = meta(0,+).
?- (:- meta_predicate(meta(X,Y))).
X = 0,
Y = +.
?-

**Comment by pmoura** [comment]
> Btw, predicate_property/2 should be able to enumerate by backtracking all
> properties of a predicate when called with an unbound second argument (the first argument is required to be bound). That will have to wait a while.

I understand. No pressure. It's not an issue for the Logtalk compiler itself.

**Comment by pmoura** [comment]
> BTW, you can currently get the meta_predicate/1 info $ cat x.pl :- meta_predicate(meta(0, +)). $ tpl x.pl Trealla Prolog (c) Infradig 2020, v1.7.20-2-g8d9c0 ?- listing. :- meta_predicate(meta(0,+)). true. ?- (:- meta_predicate(X)). X = meta(0,+). ?- (:- meta_predicate(meta(X,Y))). X = 0, Y = +. ?-
> […](#)

This will not work in general, e.g. for built-in meta-predicates. This is really information that must be available via `predicate_property/2`. One step at the time 🙂

**Comment by infradig** [comment]
What are built-in meta-predicates and how does the information get
built-in? Sorry, this stuff is new to me (although last night, by chance, I
did read ROK's rant about meta_predicate/1 and ISO Prolog, it went over my
head at the time).

**Comment by infradig** [comment]
Ok I think I see, I assume you mean definitions for the built-in functions
like call & maplist etc.

**Comment by pmoura** [comment]
> What are built-in meta-predicates and how does the information get built-in? Sorry, this stuff is new to me (although last night, by chance, I did read ROK's rant about meta_predicate/1 and ISO Prolog, it went over my head at the time).
> […](#)

I can send you a paper I wrote on meta-predicate semantics by mail. You can reach me at pmoura at logtalk dot org.

**Comment by infradig** [comment]
>Btw, predicate_property/2 should be able to enumerate by backtracking all
properties of a predicate when called with an unbound second argument (the
first argument is required to be bound).

I notice GNU Prolog allows the first arg to be unbound, allowing complete
iteration over all predicates. I have been implementing this, but maybe I
should disable it?

**Comment by pmoura** [comment]
Allowing an unbound first argument can be problematic depending on the Prolog implementation as the predicate provides access to both user and built-in predicate properties. For example, several systems support `call/N` with `N` up to maximum arity... Systems also often have a set of built-in predicates that are meant to be private and should not be exposed (a relatively common practice here is to not list any predicates that start with a `$`). In any case, the predicate specification found in the ISO Prolog standard dictates a bound first argument.

**Comment by flexoron** [comment]
Normalising expressions: 1/(2/3/(4/5))/6 -> 1/2/3/4/5/6
(Example taken from book Clause and Effect, W.F.Clocksin).

norm(X/Y,A,C) :- !, norm(X,A,B), norm(Y,B,C).
norm(X,[],X) :- !.
norm(X,A,A/X).
norm2(X,Y) :- norm(X,[],Y).

?- norm2(1/(2/3/(4/5))/6,L).
scryer:
L = 1/2/3/4/5/6.
trealla:
L = 1/2/3/4/5/6.

?- norm2(((2\*2)/2/3/(11/12)/13/(4\*5)),L).
scryer:
L = (2\*2)/2/3/11/12/13/(4\*5).
trealla:
L = 2\*2/2/3/11/12/13/4\*5.

Should be: (2\*2)...(4\*5) ?

(Trealla v1.7.44)

**Comment by infradig** [comment]
I'm going to rewrite term printing.

**Comment by flexoron** [comment]
here is an example where scryer makes a mistake? but maybe it's a problem of predicates norm2/norm:
?- norm2((3/(4/5)/6*(7/(8/9))),L).
L = ((3/(4/5)/6)*(7/(8/9))).
(4/5) and (8/9) unexpected.

**Comment by infradig** [comment]
> here is an example where scryer makes a mistake? but maybe it's a problem of predicates norm2/norm:
> ?- norm2((3/(4/5)/6*(7/(8/9))),L).
> L = ((3/(4/5)/6)*(7/(8/9))).
> (4/5) and (8/9) unexpected.

Scryer, Swipl, Gprolog & Trealla all print the same, excpet Scryer has extra parens...


$ ~/scryer-prolog/target/release/scryer-prolog x.pl
?- norm2((3/(4/5)/6*(7/(8/9))),L).
   L = (3/(4/5)/6)*(7/(8/9)).

$ tpl x.pl
?- norm2((3/(4/5)/6*(7/(8/9))),L).
L = 3/(4/5)/6*(7/(8/9)).

$ swipl x.pl
?- norm2((3/(4/5)/6*(7/(8/9))),L).
L = 3/(4/5)/6*(7/(8/9)).

$ gprolog --consult-file x.pl
| ?- norm2((3/(4/5)/6*(7/(8/9))),L).
L = 3/(4/5)/6*(7/(8/9))

**Comment by infradig** [comment]
> norm2(((2*2)/2/3/(11/12)/13/(4*5)),L).

**Comment by infradig** [comment]
> ?- norm2(((2*2)/2/3/(11/12)/13/(4*5)),L).
> scryer:
> L = (2*2)/2/3/11/12/13/(4*5).
> trealla:
> L = 2*2/2/3/11/12/13/4*5.
> 
> Should be: (2*2)...(4*5) ?
> 
> (Trealla v1.7.44)

Trealla now prints:

  $ tpl x.pl
  ?- norm2(((2*2)/2/3/(11/12)/13/(4*5)),L).
  L = 2*2/2/3/11/12/13/(4*5).

same as swipl & gprolog.

**Comment by flexoron** [comment]
that's open to dispute ... seems scryer has no doubt to print what is expected(perhaps by accident):
?- norm2(((2\*2)/2/3/(11/12)/13/(4\*5)),L).
L = (2\*2)/2/3/11/12/13/(4\*5).

**Comment by flexoron** [comment]
fyi ... mentioned Trealla here:
Scryer issue #868 Marking of operators inconsequent

**Comment by flexoron** [comment]
Trealla Prolog (c) Infradig 2020, v1.7.47
?- assertz(g(/)), assertz(g(is)), assertz(g(///)), assertz(g(iss)), g(M).
M = / 
; M = is 
; M = '///' 
; M = iss.
?- op(500,xfx,iss).
true.
?- g(M).
M = / 
; M = is 
; M = '///' 
; M = iss.

Sorry: What is iss and what is iss

**Comment by infradig** [comment]
Try v1.7.48.  Still to fix '///' in quotes though.

**Comment by flexoron** [comment]
Trealla Prolog (c) Infradig 2020, v1.7.48
?- assertz(g(/)), assertz(g(is)), assertz(g(///)), assertz(g(iss)), g(M).
M = (/) 
; M = (is) 
; M = '///' 
; M = iss.
?-  op(500,xfx,iss).
true.
?- g(M).
M = (/) 
; M = (is) 
; M = '///' 
; M = (iss).

there is nothing to fix: every atom should be displayed quoted because abc = 'abc'
the quotes allow for all possibilities.

**Comment by flexoron** [comment]
ok, something to fix
M = iss -> M = 'iss'.
M = (iss) -> M = ('iss')

**Comment by flexoron** [comment]
don't get me wrong, talking about visual output and the user who want to get what he/she see.

**Comment by flexoron** [comment]
sorry, forget the quoting issue because it is obvious now:
Trealla Prolog (c) Infradig 2020, v1.7.49
?- assertz(g(iss)), g(M).
M = iss.
?- op(500,xfx,iss), g(M).
M = (iss).

**Comment by flexoron** [comment]
btw I don't like to argue à la ada about unimportant things that much, so I report this here.
Trealla can handle it, Scryer not:
?- norm2(°a\*\*b°/2/3/(11/12)/13/4\*\*5,L).
trealla: L = '\x4294967234\\x4294967216\a'\*\*'b\x4294967234\\x4294967216\'/2/3/11/12/13/4\*\*5.
scryer: caught: error(syntax_error(unexpected_char),read_term/3:0)

who's right, who's wrong?

**Comment by flexoron** [comment]
Trealla:
?- norm2(°2**2°/2/3/(11/12)/13/4**5,L).
false.

**Comment by infradig** [comment]
Trealla allows any UTF8 character that's not space, control, digit or
special character to start an atom. I don't know how this jibes with the
ISO Prolog standard. I will look into it.

**Comment by infradig** [comment]
For example...

?- S = 一二三.
S = 一二三.

in both tpl & swipl.

**Comment by flexoron** [comment]
?- X is 4 \* 2°.
X = 8.

**Comment by infradig** [comment]
Interesting. I've been working on utf8 printing, I guess I need to look at
parsing.

**Comment by infradig** [comment]
Ok, I think I have sorted out the utf-8 alpha stuff.

**Comment by flexoron** [comment]
Trealla Prolog (c) Infradig 2020, v1.7.55
?- X is 4 \* 2°.
Error: syntax error, missing operand or operator
OK

?- X is \*(4°,2).
uncaught exception: error(type_error(evaluable,'°'/0),(*)/2)
Why uncaught exception and not 'syntax error', too?

?- X is \*(4,2°).
X = 8.
Not ok

**Comment by infradig** [comment]
> Trealla Prolog (c) Infradig 2020, v1.7.55
> ?- X is 4 * 2°.
> Error: syntax error, missing operand or operator
> OK

Because is a digit followed by an alpha IS a syntax error (which can only
occur during parsing).

> ?- X is *(4°,2).
> uncaught exception: error(type_error(evaluable,'°'/0),(*)/2)
> Why uncaught exception and not 'syntax error', too?

The parser doesn't know a number is expected at this place, it is only
detected at run time.

> ?- X is *(4,2°).
> X = 8.
> Not ok

Definitely got me there! 😎

**Comment by infradig** [comment]
Ignore what I just said. The last two are definitely wrong.

**Comment by flexoron** [comment]
very nice but an issue with full stop

Trealla Prolog (c) Infradig 2020, v1.7.55
?- [user].
°(C) :- F is (C \* (9/5)) + 32,
          write(C), write('°C = '), write(F), write('°F'), nl.
^D
?- op(500,yf,°).

?- 2° .
2°C = 35.6°F

?- 2°.
Error: syntax error, missing operand or operator
true.

**Comment by infradig** [comment]
> ?- 2° .
2°C = 35.6°F

What's between degree and dot?

**Comment by infradig** [comment]
I get this with a space...

$ tpl
Trealla Prolog (c) Infradig 2020, v1.7.55-1-g586bd
?- op(500,yf,°).
true.
?- 2° .
uncaught exception: error(existence_error(procedure,2°),(°)/1)
false.

**Comment by infradig** [comment]
A man, I forgot to input the predicate. I need to stop for the day.

**Comment by infradig** [comment]
> ?- 2°.

> Error: syntax error, missing operand or operator
> true.

The parser is joining the degree & the dot as one atom. I will fix.

**Comment by infradig** [comment]
ok, fixed this one.

**Comment by pmoura** [comment]
On macOS:

```text
cc -Isrc -I/usr/local/include -DVERSION='"v1.7.55-4-g0498f"' -O3  -Wall -Wextra -D_GNU_SOURCE   -c -o src/predicates.o src/predicates.c
src/predicates.c:4176:2: warning: expression which evaluates to zero treated as a null pointer constant of type 'cell *'
      (aka 'struct cell_ *') [-Wnon-literal-null-conversion]
        may_ptr_error(init_tmp_heap(q));
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/internal.h:203:81: note: expanded from macro 'may_ptr_error'
#define may_ptr_error(expr, ...) CHECK_SENTINEL(expr, NULL, __VA_ARGS__; return pl_error)
                                                                                ^~~~~~~~
src/internal.h:198:90: note: expanded from macro 'CHECK_SENTINEL'
#define CHECK_SENTINEL(expr, err_sentinel, ...) CHECK_SENTINEL_((expr), err_sentinel, ## __VA_ARGS__, error=true)
                                                                                         ^~~~~~~~~~~
src/internal.h:199:128: note: expanded from macro 'CHECK_SENTINEL_'
  ...err_sentinel, on_error, ...) do { if((expr) == err_sentinel){message(#expr " = " #err_sentinel); on_error;}} while (0)
                                                                                                      ^~~~~~~~
1 warning generated.
```

**Comment by infradig** [comment]
Thx. Only clang picked that up.

**Comment by pmoura** [comment]
Thanks for the quick fix. Get some sleep! 🙃

**Comment by flexoron** [comment]
neat, fixed, thnx.
?- [user].
°(X) :-
   F is (X \* (9/5)) + 32,
   write(X), write('°C = '), write(F), write('°F'), nl,
   C is (X - 32) \* 5/9,
   write(X), write('°F = '), write(C), write('°C'), nl.

?- op(500,yf,°).
?- 36°.
36°C = 96.8°F
36°F = 2.2222222222222223°C

**Comment by flexoron** [comment]
?- (36,37)°.
Segmentation fault (core dumped)

**Comment by flexoron** [comment]
Cool! lets continue core dump hunting a little.

?- [user].
°(X) :-
   F is (X \* (9/5)) + 32,
   write(X), write('°C = '), write(F), write('°F'), nl,
   C is (X - 32) \* 5/9,
   write(X), write('°F = '), write(C), write('°C'), nl.

°°([C,F]) :-
    var(F), var(C), C = (F-32)\*5/9, F = (C\*(9/5))+32, !;
    var(F), F is (C \* (9/5)) + 32, !;
    var(C), C is (F - 32) \* 5 / 9, !;
    X is (C \* (9/5)) + 32, float(X) == float(F).

?- op(500,yf,°).
?- op(500,yf,°°).

?- C = 36, F = 96.8, [C,F]°°.
C = 36,
F = 96.8.

?- C = 36, F = 96, [C,F]°°.
false.

?- C = 36,  [C,F]°°.
C = 36,
F = 96.8.

?- C = 36, F = 96.8, [C,F]°°, C°.
36°C = 96.8°F
36°F = 2.2222222222222223°C
C = 36,
F = 96.8.

?- C = 36, F = 96.8, [C,F]°°, C°, F°.
36°C = 96.8°F
36°F = 2.2222222222222223°C
96.8°C = 206.24000000000001°F
96.8°F = 36.0°C
C = 36,
F = 96.8.

?- [C,F]°°.
C = (F-32)\*5/9,
F = C\*(9/5)+32.

?- [C,F]°°, C°, F°.
Segmentation fault (core dumped)

**Comment by infradig** [comment]
Not seeing this, maybe got fixed with that last commit.

**Comment by flexoron** [comment]
nope, just cloned 
Trealla Prolog (c) Infradig 2020, v1.7.55-6-gf553b
?- [C,F]°°, C°.
Segmentation fault (core dumped)

**Comment by flexoron** [comment]
git log
commit f553b9b5c91f4c7a21fc25351f5f655d83ff5389
Date:   Mon Mar 15 23:09:02 2021 +1000

**Comment by flexoron** [comment]
and here I did not define op °, no op(500,yf,°).

?-  [C,F]°°, C°.
src/predicates.c:5109 throw_error: len >= 0 failed 
Aborted (core dumped)

**Comment by infradig** [comment]
I'm not seeing any of these errors. I'll keep an eye on it.

**Comment by flexoron** [comment]
may I ask what you see: 
?- [C,F]°°, C°.

**Comment by josd** [comment]
Here I see:
```
$ tpl
Trealla Prolog (c) Infradig 2020, v1.7.55-6-gf553b
?- [C,F]°°, C°.
Error: syntax error, operator expected '.', line nbr 0
false.
?-
```

**Comment by infradig** [comment]
$ tpl
Trealla Prolog (c) Infradig 2020, v1.7.55-6-gf553b
?- [C,F]°°, C°.
Error: syntax error, operator expected '.', line nbr 0
false.
?-
***@***.***:~/trealla$

**Comment by flexoron** [comment]
Trealla Prolog (c) Infradig 2020, v1.7.55-6-gf553b
?- [user].
°(X) :-
F is (X \* (9/5)) + 32,
write(X), write('°C = '), write(F), write('°F'), nl,
C is (X - 32) \* 5/9,
write(X), write('°F = '), write(C), write('°C'), nl.

°°([C,F]) :-
var(F), var(C), C = (F-32)\*5/9, F = (C\*(9/5))+32, !;
var(F), F is (C \* (9/5)) + 32, !;
var(C), C is (F - 32) \* 5 / 9, !;
X is (C \* (9/5)) + 32, float(X) == float(F).
true.
?- listing.
'°°'([C,F]) :- var(F),var(C),C=(F-32)\*5/9,F=C*(9/5)+32,!;var(F),F is C\*(9/5)+32,!;var(C),C is (F-32)\*5/9,!;X is C*(9/5)+32,float(X)==float(F).
'°'(X) :- F is X\*(9/5)+32,write(X),write('°C = '),write(F),write('°F'),nl,C is (X-32)\*5/9,write(X),write('°F = '),write(C),write('°C'),nl.
true.
?- op(500,yf,°°).
true.
?- op(500,yf,°).
true.
?- [C,F]°°, C°.
Segmentation fault (core dumped)

**Comment by flexoron** [comment]
what you see here:
?- [C,F]°°.

**Comment by flexoron** [comment]
whats that (nothing defined, no predicates, no ops)

Trealla Prolog (c) Infradig 2020, v1.7.55-6-gf553b
?- [C,F]°°, C°.
Error: syntax error, operator expected '.', line nbr 0
false.
?- [C,F]°°.
true.

**Comment by infradig** [comment]
> ?- [C,F]°°.
> true.

Ok, I get that one. For the to-do list.

**Comment by flexoron** [comment]
you guys have not defined the predicates ° and °°.

**Comment by infradig** [comment]
Your example didn't. But it makes no difference.

**Comment by infradig** [comment]
Actually, I made a mistake in my op definition. Now I see it!!!

**Comment by flexoron** [comment]
both?

Trealla Prolog (c) Infradig 2020, v1.7.55-6-gf553b

?- op(500,yf,°°).

?- [C,F]°°, C°.
src/predicates.c:5109 throw_error: len >= 0 failed 
Aborted (core dumped)

++++++++++++++++++++++++++++++++++++++
Trealla Prolog (c) Infradig 2020, v1.7.55-6-gf553b

?- op(500,yf,°).
?- op(500,yf,°°).

?- [C,F]°°, C°.
Segmentation fault (core dumped)

**Comment by infradig** [comment]
Yes. Have to have both calls for it to happen. Probably the op is grossly
distorting things...

  ?-  ([C,F]°°), C°

doesn't crash..

**Comment by flexoron** [comment]
here it does crash:
Case 1:
?- op(500,yf,°).
?- op(500,yf,°°).
?-  ([C,F]°°), C°.
Segmentation fault (core dumped)

Case 2:
?- op(500,yf,°°).
?- ([C,F]°°), C°.
src/predicates.c:5109 throw_error: len >= 0 failed 
Aborted (core dumped)

**Comment by josd** [comment]
When taking a higher operator priority it just gives errors:
```
$ tpl
Trealla Prolog (c) Infradig 2020, v1.7.55-6-gf553b
?- [user].
°(X) :-
F is (X * (9/5)) + 32,
write(X), write('°C = '), write(F), write('°F'), nl,
C is (X - 32) * 5/9,
write(X), write('°F = '), write(C), write('°C'), nl.

°°([C,F]) :-
var(F), var(C), C = (F-32)*5/9, F = (C*(9/5))+32, !;
var(F), F is (C * (9/5)) + 32, !;
var(C), C is (F - 32) * 5 / 9, !;
X is (C * (9/5)) + 32, float(X) == float(F).
true.
?- op(1150,yf,°).
true.
?- op(1150,yf,°°).
true.
?- ([C,F]°°), C°.
uncaught exception: error(type_error(evaluable,alse/0),(*)/2)
false.
?- [C,F]°°, C°.
Error: syntax error, operator expected '.', line nbr 0
false.
?-
```

**Comment by josd** [comment]
and seems to work fine
```
?- 8°.
8°C = 46.4°F
8°F = -13.333333333333334°C
true.
?- [8,X]°°.
X = 46.4.
?-
```

**Comment by infradig** [comment]
Yes, individually both work fine, But in conjunction there's a problem.

**Comment by flexoron** [comment]
higher operator priority...true...interesting but is it relevant?
zero priority works, too.
?- op(0,yf,°).
?- op(0,yf,°°).
?- ([C,F]°°), C°.
Error: syntax error, operator expected '.', line nbr 0

**Comment by josd** [comment]
Priority 950 seems fine to work with the conjunction:
```
$ tpl
Trealla Prolog (c) Infradig 2020, v1.7.55-6-gf553b
?- [user].
°(X) :-
F is (X * (9/5)) + 32,
write(X), write('°C = '), write(F), write('°F'), nl,
C is (X - 32) * 5/9,
write(X), write('°F = '), write(C), write('°C'), nl.

°°([C,F]) :-
var(F), var(C), C = (F-32)*5/9, F = (C*(9/5))+32, !;
var(F), F is (C * (9/5)) + 32, !;
var(C), C is (F - 32) * 5 / 9, !;
X is (C * (9/5)) + 32, float(X) == float(F).
true.
?- op(950,yf,°).
true.
?- op(950,yf,°°).
true.
?- [C,F]°°, 8°.
8°C = 46.4°F
8°F = -13.333333333333334°C
C = (F-32)*5/9,
F = C*(9/5)+32.
?-
```

**Comment by flexoron** [comment]
you see what C and F are: C need F and F need C...this 'return' is not good for for C° or F°.
?- [C,F]°°.
C = (F-32)*5/9,
F = C*(9/5)+32.

**Comment by josd** [comment]
It is not wrong what it says and swipl even gives:
```
?- ([C,F]°°).
C =  (C*(9/5)+32-32)*5/9,
F = C*(9/5)+32.
```

**Comment by flexoron** [comment]
@infradig the priority must be 0 or higher than  (',').
?- current_op(A, _, (',')).
A = 1000,

?- op(1001,yf,°).
?- op(1001,yf,°°).

@josd ok maybe  but unclear.
the definition clearly says (F-32) and (C\*(9/5))
°°([C,F]) :-
var(F), var(C), C = (F-32)*5/9, F = (C\*(9/5))+32, !;

Trealla is not wrong and does it much better than others. (I get what I expect, thx trealla)

**Comment by infradig** [comment]
op/3 with priority 0 means delete the operator.

**Comment by flexoron** [comment]
ups ok
priority 1001 avoids the core dumps

**Comment by flexoron** [comment]
@josd , the parenthesis are non-desirable in this context.
swipl:
?- [C,F]°°.    .
ERROR: Syntax error ...
?- [C,F]°° .
C =  (C\*(9/5)+32-32)\*5/9,
F = C\*(9/5)+32.

swipl has an issue with full stop, trealla fixed this already(see above).
swipl confusing the user(me in this case):
C =  (C*(9/5)+32-32)\*5/9,

**Comment by infradig** [comment]
Hey Paulo, I'm thinking of switching to using *xxd* to convert library .pl
files to binary objects. I used to use it once before but it wasn't
available on FreeBSD (or so I thought, turns out it is in the the
editors/vim package). If it is available on macOS (and something I read
tells me it is) then I can use that and hey presto, no more NOLIBS flag
neededfor Macs.

**Comment by pmoura** [comment]
> Hey Paulo, I'm thinking of switching to using *xxd* to convert library .pl files to binary objects. I used to use it once before but it wasn't available on FreeBSD (or so I thought, turns out it is in the the editors/vim package). If it is available on macOS (and something I read tells me it is) then I can use that and hey presto, no more NOLIBS flag neededfor Macs.
> […](#)

Do you mean this:
https://ss64.com/osx/xxd.html

**Comment by infradig** [comment]
Yes, that's it.

**Comment by infradig** [comment]
I have pushed changes to branch *xxd*. When you have a chance let me know
how it goes on macOS, It works on Linux and FreeBSD.

**Comment by pmoura** [comment]
> I have pushed changes to branch *xxd*. When you have a chance let me know how it goes on macOS, It works on Linux and FreeBSD.
> […](#)

Switching to the `xxd` branch and building on macOS 11.2.3 using `gmake clean && gmake NOSSL=1` works. I get a fully functional build of `tpl` on my limited experiments.

**Comment by infradig** [comment]
Great! Thanks.

**Comment by flexoron** [comment]
Trealla Prolog (c) Infradig 2020, v1.7.59
?- op(500,yf,+++).
true.
?- op(0,yf,+++).
uncaught exception: error(type_error(atom,yf+++),op/3)
false.

it should be allowed to delete ops ?

**Comment by infradig** [comment]
Hmm, op(0,yf,(+++)) works, so parser shenanigans. I wrote an experimental
new parser for Trealla last year, but never got around to seeing it
through. It's still in the toy stage, but if I find time i'd like to give
it a try out. In the meantime I can try some workarounds.

**Comment by flexoron** [comment]
its ok .. I stop tantalise the parser for a while.
Trealla Prolog (c) Infradig 2020, v1.7.59
?- op(500,yf,(+++)).
true.
?- current_op(A, _, (+++)).
false.

**Comment by infradig** [comment]
Pushed a fix for this one...

$ tpl
?- op(500,yf,+++).
true.
?- current_op(P,S,+++).
P = 500,
S = yf
; false.
?- op(0,yf,+++).
true.
?- current_op(P,S,+++).
false.

**Comment by flexoron** [comment]
I'm afraid I made  a mistake.
It is unclear whether postfix op ° and  full stop need a space in between.
(see scryer closed issue: Postfix operator syntax, Parser error #877)

and

scryer, gprolog, swipl:
?- atom(+.).
true.

$ ./tpl
Trealla Prolog (c) Infradig 2020, v1.7.59
?- atom(+.).
false.

**Comment by flexoron** [comment]
next issue with atom()
(see scryer closed issue: A question about atom()  #880)

Scryer:
?- atom(+.)
true
?- atom(a.).
caught: error(syntax_error(incomplete_reduction),read_term/3:1)

mthom: I think this is because + is a graphic token and a is alphabetic.
 
Trealla Prolog (c) Infradig 2020, v1.7.70
?- atom(+.)
true
?- atom(a.)
true.

**Comment by infradig** [comment]
Correct, i'll look into it...
